pytorch_lightning/loops/optimization/optimizer_loop.py

```diff
@@ -226,6 +226,7 @@ def advance(self, batch: Any, *args: Any, **kwargs: Any) -> None:  # type: ignor
         self.optim_progress.optimizer_position += 1
 
     def on_run_end(self) -> _OUTPUTS_TYPE:
+        self.optim_progress.optimizer_position = 0
```

Comment:
@carmocca you commented on a PR here #9371 that you prefer the reset logic for the optim progress to live inside the `reset()` method entirely. But this doesn't work anymore, because the above `self.done` call inside reset is always True due to `self._indices` being an empty list. The list cannot be populated until `on_run_start`. 

The alternative is to override run:

```py
    def run(self, batch: Any, optimizers: List[Tuple[int, Optimizer]], batch_idx: int):
        self._indices, self._optimizers = zip(*optimizers)
        return super().run(batch, optimizers, batch_idx)
```
to populate the list as early as possible so `reset()` can run `self.done`. Would you prefer this or the current PR content?

Reply:
The current solution looks simpler, but I see a third option which is to save the `self._indices` as part of the tracking so that we can still call `self.done`

Reply:
I think overriding run seems the best option no ? It means we could move BaseProgress reset_on_restart within the reset function.

Reply:
@carmocca and I found a shorter solution that does not require overriding `run()` or move the existing code for indices.